### PR TITLE
fix errors in Either 'app' and Apply 'applyFirst'

### PR DESCRIPTION
**Pull request by sledorze**
I've based the additional tests from results of Purescript implementations and fixed the code accordingly.

### CODE REVIEW

**Comment by gcanti** [comment]
@sledorze two nitpicks before merging: I think we could use `constant` instead of defining a new function every time `applyFirst` is called. Perhaps we could also extract `() => identity` from `applySecond`

```ts
const constIdentity = () => identity

export function applySecond<F extends HKT2S>(apply: StaticApply<F>): <L, A, B>(fa: HKT2<L, A>[F], fb: HKT2<L, B>[F]) => HKT2<L, B>[F]
export function applySecond<F extends HKTS>(apply: StaticApply<F>): <A, B>(fa: HKT<A>[F], fb: HKT<B>[F]) => HKT<B>[F]
export function applySecond<F extends HKTS>(apply: StaticApply<F>): <A, B>(fa: HKT<A>[F], fb: HKT<B>[F]) => HKT<B>[F] {
  return <A, B>(fa: HKT<A>[F], fb: HKT<B>[F]) => apply.ap(apply.map(constIdentity, fa), fb)
}
```

What do you think?

**Comment by gcanti** [comment]
However we can do it later, I suspect I left around other micro-optimizations like this, for example here https://github.com/gcanti/fp-ts/blob/master/src/Monoid.ts#L25

File: src/Apply.ts [hunk 1]
```diff
@@ -13,13 +13,13 @@ export interface FantasyApply<F extends HKTS, A> extends FantasyFunctor<F, A> {
...
 export function applySecond<F extends HKTS>(apply: StaticApply<F>): <A, B>(fa: HKT<A>[F], fb: HKT<B>[F]) => HKT<B>[F] {
-  return <A, B>(fa: HKT<A>[F], fb: HKT<B>[F]) => apply.ap<B, B>(apply.map(() => identity, fa), fb)
+  return <A, B>(fa: HKT<A>[F], fb: HKT<B>[F]) => apply.ap<B, B>(apply.map<A, (x: B) => B>((_: A) => identity, fa), fb)
 }
 
 export function applyFirst<F extends HKT2S>(apply: StaticApply<F>): <L, A, B>(fa: HKT2<L, A>[F], fb: HKT2<L, B>[F]) => HKT2<L, A>[F]
 export function applyFirst<F extends HKTS>(apply: StaticApply<F>): <A, B>(fa: HKT<A>[F], fb: HKT<B>[F]) => HKT<A>[F]
 export function applyFirst<F extends HKTS>(apply: StaticApply<F>): <A, B>(fa: HKT<A>[F], fb: HKT<B>[F]) => HKT<A>[F] {
-  return <A, B>(fa: HKT<A>[F], fb: HKT<B>[F]) => apply.ap<A, A>(apply.map(() => identity, fb), fa)
+  return <A, B>(fa: HKT<A>[F], fb: HKT<B>[F]) => apply.ap<B, A>(apply.map<A, (x: B) => A>((a: A) => (b: B) => a, fa), fb)
```
Comment (sledorze) [inline src/Apply.ts]: oh.. that one :)

File: src/Apply.ts [hunk 2]
```diff
@@ -13,13 +13,13 @@ export interface FantasyApply<F extends HKTS, A> extends FantasyFunctor<F, A> {
 export function applySecond<F extends HKT2S>(apply: StaticApply<F>): <L, A, B>(fa: HKT2<L, A>[F], fb: HKT2<L, B>[F]) => HKT2<L, B>[F]
 export function applySecond<F extends HKTS>(apply: StaticApply<F>): <A, B>(fa: HKT<A>[F], fb: HKT<B>[F]) => HKT<B>[F]
 export function applySecond<F extends HKTS>(apply: StaticApply<F>): <A, B>(fa: HKT<A>[F], fb: HKT<B>[F]) => HKT<B>[F] {
-  return <A, B>(fa: HKT<A>[F], fb: HKT<B>[F]) => apply.ap<B, B>(apply.map(() => identity, fa), fb)
+  return <A, B>(fa: HKT<A>[F], fb: HKT<B>[F]) => apply.ap<B, B>(apply.map<A, (x: B) => B>((_: A) => identity, fa), fb)
```
Comment (sledorze) [inline src/Apply.ts]: @gcanti I am not sure if we should help or not the compiler (and curious reader) with type annotations (compiler, for speed).

**Comment by sledorze** [comment]
@gcanti I prefer picky code reviews like that rather than "let's do it later.." :)

**Comment by gcanti** [comment]
Ok, great! ;)

**Pull request merged**